pkg/share/manager/cs3/cs3.go

```diff
@@ -0,0 +1,564 @@
+// Copyright 2018-2022 CERN
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// In applying this license, CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+package cs3
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"path"
+	"time"
+
+	groupv1beta1 "github.com/cs3org/go-cs3apis/cs3/identity/group/v1beta1"
+	userpb "github.com/cs3org/go-cs3apis/cs3/identity/user/v1beta1"
+	collaboration "github.com/cs3org/go-cs3apis/cs3/sharing/collaboration/v1beta1"
+	provider "github.com/cs3org/go-cs3apis/cs3/storage/provider/v1beta1"
+	typespb "github.com/cs3org/go-cs3apis/cs3/types/v1beta1"
+	ctxpkg "github.com/cs3org/reva/v2/pkg/ctx"
+	"github.com/cs3org/reva/v2/pkg/errtypes"
+	"github.com/cs3org/reva/v2/pkg/share"
+	"github.com/cs3org/reva/v2/pkg/share/manager/registry"
+	"github.com/cs3org/reva/v2/pkg/storage/utils/indexer"
+	"github.com/cs3org/reva/v2/pkg/storage/utils/indexer/option"
+	"github.com/cs3org/reva/v2/pkg/storage/utils/metadata"
+	"github.com/cs3org/reva/v2/pkg/utils"
+	"github.com/google/uuid"
+	"github.com/mitchellh/mapstructure"
+	"github.com/pkg/errors"
+	"google.golang.org/genproto/protobuf/field_mask"
+)
+
+//go:generate mockery -name Storage
+//go:generate mockery -name Indexer
+
+// Storage is the interface to the metadata storage backend
+type Storage interface {
+	metadata.Storage
+}
+
+// Indexer is the interface to the indexer being used for indexing shares
+type Indexer interface {
+	AddIndex(t interface{}, indexBy option.IndexBy, pkName, entityDirName, indexType string, bound *option.Bound, caseInsensitive bool) error
+	Add(t interface{}) ([]indexer.IdxAddResult, error)
+	FindBy(t interface{}, field string, val string) ([]string, error)
+	Delete(t interface{}) error
+}
+
+// Manager implements a share manager using a cs3 storage backend
+type Manager struct {
+	storage Storage
+	indexer Indexer
+
+	initialized bool
+}
+
+// ReceivedShareMetadata hold the state information or a received share
+type ReceivedShareMetadata struct {
+	State      collaboration.ShareState `json:"state"`
+	MountPoint *provider.Reference      `json:"mountpoint"`
+}
+
+func init() {
+	registry.Register("cs3", NewDefault)
+}
+
+type config struct {
```

Comment:
Can we make this public? Having an exported overview of the configuration for this might be beneficial

Reply:
Hm, maybe. But currently the struct is only used for extracting the configuration values from the generic interface which is being passed as part of the registry contract. The values are only passed on to the storage and not even stored in the manager itself.
Personally I would rather keep it private and expose it when it's actually being used.